torch_dicom/transforms/overlay.py

```diff
@@ -0,0 +1,228 @@
+                unit="image",
+                disable=not use_bar,
+                leave=False,
+            )
+        )
+        self.p = p
+        self.scale_range = scale
+        self.force_in_bounds = force_in_bounds
+
+        # TODO: Implement this
+        if not self.force_in_bounds:
+            raise NotImplementedError("`force_in_bounds=False` is not yet implemented")  # pragma: no cover
+
+    def _transform(self, inpt: Any, params: Dict[str, Any]) -> Any:
+        if not params["augment"] or not isinstance(inpt, TVImage):
+            return inpt
+        if inpt.ndim > 3:
+            raise ValueError("Batched inputs are not supported")
+
+        # Load inputs
+        idx = params["watermark_idx"]
+        scale = params["scale"]
+        center = params["center"]
+        H, W = inpt.shape[-2:]
+        watermark = self._load_image(idx).to(inpt.device)
+
+        # Rescale the overlay according to random scale factor.
+        # Ensure that the overlay is not larger than the input image.
+        Hw, Ww = watermark.shape[-2:]
+        Hw_scale, Ww_scale = int(Hw * scale), int(Ww * scale)
+        Hw_scale = min(Hw_scale, H)
+        Ww_scale = min(Ww_scale, W)
+        watermark = F.interpolate(
+            watermark.view(1, 2, Hw, Ww),
+            size=(Hw_scale, Ww_scale),
+            mode="nearest",
+        ).view(2, Hw_scale, Ww_scale)
+
+        # Determine bounds for indexing into the input image
+        Hw, Ww = watermark.shape[-2:]
+        lower_bounds = torch.stack(
+            [
+                center[..., 0] - Hw // 2,
+                center[..., 1] - Ww // 2,
+            ],
+            dim=-1,
+        )
+        # If we require the entire overlay to be in bounds, we need to check
+        # that the randomly chosen center is not too close to the edge. If it
+        # is, we need to choose a new center.
+        if self.force_in_bounds:
+            lower_bounds.clip_(min=0)
+            lower_bounds = torch.min(lower_bounds, lower_bounds.new_tensor([H - Hw, W - Ww]))
+
+        # Otherwise, if we are out of bounds we need to slice the overlay
+        else:
+            # TODO: Implement this
+            raise NotImplementedError  # pragma: no cover
+
+        # Determine the upper bounds based on watermark size and lower bounds
+        upper_bounds = lower_bounds + lower_bounds.new_tensor([Hw, Ww])
+        bounds = torch.cat([lower_bounds, upper_bounds], -1)
+        assert (
+            (bounds[..., 0] >= 0).all()
+            and (bounds[..., 1] >= 0).all()
+            and (bounds[..., 2] <= H).all()
+            and (bounds[..., 3] <= W).all()
+        ), f"Bounds {bounds} are out of range for image of size {H}x{W}"
+
+        # Apply the watermark
+        target_for_watermark = inpt[..., bounds[0] : bounds[2], bounds[1] : bounds[3]]
+        watermark, alpha = watermark.split(1)
+        target_for_watermark = torch.where(alpha > 0, watermark, target_for_watermark)
```

Comment:
Since there is no alpha blending with transparency, should we assert that alpha values are either 0 or 1? (or maybe just issue a warning if alpha values greater than 0 but less than 1 are used)

Reply:
I updated the code to apply proper alpha blending with transparency